OpenRA.Mods.Common/FileFormats/WavReader.cs

```diff
@@ -112,69 +117,86 @@ public static float WaveLength(Stream s)
 			return length / (channels * sampleRate * bitsPerSample);
 		}
 
-		public byte[] DecodeImaAdpcmData()
+		sealed class WavStream : ReadOnlyAdapterStream
 		{
-			var s = new MemoryStream(RawOutput);
-
-			var numBlocks = DataSize / BlockAlign;
-			var blockDataSize = BlockAlign - (Channels * 4);
-			var outputSize = UncompressedSize * Channels * 2;
-
-			var outOffset = 0;
-			var output = new byte[outputSize];
-
-			var predictor = new int[Channels];
-			var index = new int[Channels];
+			readonly short channels;
+			readonly int numBlocks;
+			readonly int blockDataSize;
+			readonly int outputSize;
+			readonly int[] predictor;
+			readonly int[] index;
+
+			readonly byte[] interleaveBuffer;
+			int outOffset;
+			int currentBlock;
+
+			public WavStream(Stream stream, int dataSize, short blockAlign, short channels, int uncompressedSize) : base(stream)
+			{
+				this.channels = channels;
+				numBlocks = dataSize / blockAlign;
+				blockDataSize = blockAlign - (channels * 4);
+				outputSize = uncompressedSize * channels * 2;
+				predictor = new int[channels];
+				index = new int[channels];
+
+				interleaveBuffer = new byte[channels * 16];
+			}
 
-			// Decode each block of IMA ADPCM data in RawOutput
-			for (var block = 0; block < numBlocks; block++)
+			protected override bool BufferData(Stream baseStream, Queue<byte> data)
 			{
+				// Decode each block of IMA ADPCM data
 				// Each block starts with a initial state per-channel
-				for (var c = 0; c < Channels; c++)
+				for (var c = 0; c < channels; c++)
 				{
-					predictor[c] = s.ReadInt16();
-					index[c] = s.ReadUInt8();
-					/* unknown/reserved */ s.ReadUInt8();
+					predictor[c] = baseStream.ReadInt16();
+					index[c] = baseStream.ReadUInt8();
+					baseStream.ReadUInt8(); // Unknown/Reserved
 
 					// Output first sample from input
-					output[outOffset++] = (byte)predictor[c];
-					output[outOffset++] = (byte)(predictor[c] >> 8);
+					data.Enqueue((byte)predictor[c]);
+					data.Enqueue((byte)(predictor[c] >> 8));
+					outOffset += 2;
 
 					if (outOffset >= outputSize)
-						return output;
+						return true;
 				}
 
 				// Decode and output remaining data in this block
 				var blockOffset = 0;
 				while (blockOffset < blockDataSize)
 				{
-					for (var c = 0; c < Channels; c++)
+					for (var c = 0; c < channels; c++)
 					{
 						// Decode 4 bytes (to 16 bytes of output) per channel
-						var chunk = s.ReadBytes(4);
+						var chunk = baseStream.ReadBytes(4);
 						var decoded = ImaAdpcmReader.LoadImaAdpcmSound(chunk, ref index[c], ref predictor[c]);
 
 						// Interleave output, one sample per channel
-						var outOffsetChannel = outOffset + (2 * c);
+						var interleaveChannelOffset = 2 * c;
 						for (var i = 0; i < decoded.Length; i += 2)
 						{
-							var outOffsetSample = outOffsetChannel + i;
-							if (outOffsetSample >= outputSize)
-								return output;
```

Comment:
Removing this was intentional, we no longer bail out whilst in the middle of decoding each channel.